## Fundamentos de programación en Python

**Propedéutico — FLACSO Ecuador** · Duración: 2 horas · **VERSIÓN DOCENTE (solucionario ejecutable)**

> Esta versión contiene las soluciones de todos los ejercicios como **celdas de código visibles y ejecutables**: permite proyectar en clase, correr **Run All** y verificar resultados. La versión para estudiantes mantiene las soluciones ocultas en bloques desplegables.

| Bloque | Tema | Tiempo |
|---|---|---|
| 0 | Verificación del entorno | 5 min |
| 1 | Algoritmos: la idea central | 10 min |
| 2 | Variables y tipos de datos | 20 min |
| 3 | Operadores | 15 min |
| 4 | Estructuras de datos: listas, tuplas y diccionarios | 20 min |
| 5 | Estructuras de control: `if`, `for`, `while` | 25 min |
| 6 | Funciones | 15 min |
| 7 | Ejercicio integrador | 10 min |

**Objetivos.** Al finalizar la sesión podrás: (1) describir un problema como un algoritmo, (2) usar variables, tipos y operadores de Python, (3) almacenar información en listas y diccionarios, (4) controlar el flujo de un programa con `if`, `for` y `while`, y (5) encapsular lógica en funciones reutilizables.

**Metodología.** Cada bloque sigue el patrón **🧑‍🏫 Yo hago** (demostración guiada) → **🔨 Tú haces** (ejercicio individual con solución oculta). En esta versión docente cada ejercicio va seguido de su **✅ Solución** ejecutable.

## 0. Verificación del entorno (5 min)

Cómo trabajar en VS Code con este notebook:

1. Abre la carpeta del curso: **Archivo → Abrir carpeta** y coloca ahí este archivo `.ipynb`.
2. Selecciona el *kernel* de Python: botón **Select Kernel** (esquina superior derecha) → tu instalación de Python 3.
3. **Shift + Enter** ejecuta la celda actual y avanza a la siguiente. **Ctrl + Enter** ejecuta sin avanzar.
4. Las celdas se ejecutan **en orden, de arriba hacia abajo**: si saltas celdas, algunas variables no existirán todavía.

Ejecuta la celda siguiente. Si ves el saludo y la versión de Python, tu entorno está listo.

In [1]:
# Verificación del entorno
import sys

print("¡Hola, propedéutico FLACSO!")
print("Versión de Python:", sys.version.split()[0])

¡Hola, propedéutico FLACSO!
Versión de Python: 3.12.3


## 1. Algoritmos: la idea central (10 min)

Un **algoritmo** es una secuencia **finita y ordenada de pasos precisos** que transforma **entradas** en **salidas**. Una receta de cocina es un algoritmo: ingredientes (entrada) → pasos (proceso) → plato (salida).

Todo algoritmo bien planteado cumple cuatro propiedades:

| Propiedad | Significado |
|---|---|
| Preciso | Cada paso está definido sin ambigüedad |
| Finito | Termina en un número limitado de pasos |
| Ordenado | El orden de los pasos importa |
| Efectivo | Produce el resultado esperado |

Antes de escribir código conviene pensar el algoritmo en **pseudocódigo** (lenguaje natural estructurado):

```
Entrada:  tres calificaciones
Proceso:  sumarlas y dividir el total para 3
Salida:   el promedio
```

Programar es, esencialmente, **traducir algoritmos a un lenguaje que la computadora entienda**. Veamos esa traducción a Python.

In [2]:
# 🧑‍🏫 Yo hago — Algoritmo: promedio de tres calificaciones
# Entrada:  tres calificaciones
# Proceso:  sumarlas y dividir el total para 3
# Salida:   el promedio

nota_1 = 8.5
nota_2 = 9.0
nota_3 = 7.5

suma = nota_1 + nota_2 + nota_3
promedio = suma / 3

print("El promedio es:", promedio)

El promedio es: 8.333333333333334


### 🔨 Tú haces 1

En la celda siguiente, escribe **solo en comentarios** (líneas que empiezan con `#`) el algoritmo para decidir si un número entero es **par o impar**. Pista: piensa en el **residuo** de dividirlo entre 2. Lo implementaremos en código en el bloque 3.

In [3]:
# Tu algoritmo aquí (solo comentarios):
# Entrada:
# Proceso:
# Salida:

#### ✅ Solución

In [4]:
# Entrada:  un número entero n
# Proceso:  calcular el residuo de n ÷ 2
#           si el residuo es 0  → el número es par
#           si el residuo es 1  → el número es impar
# Salida:   el mensaje "par" o "impar"

## 2. Variables y tipos de datos (20 min)

Una **variable** es un nombre que guarda un valor. Se crea con el operador de asignación `=` (leer como "toma el valor de", no como igualdad matemática).

Reglas para nombrar variables:

| Regla | ✅ Correcto | ❌ Incorrecto |
|---|---|---|
| Sin espacios: usar `snake_case` | `tasa_interes` | `tasa interes` |
| No empezar con número | `anio_2026` | `2026_anio` |
| Sin tildes ni ñ (recomendado) | `anio` | `año` |
| Distingue mayúsculas | `pib` y `PIB` son variables distintas | — |

Los cuatro **tipos básicos**:

| Tipo | Nombre | Ejemplo |
|---|---|---|
| `int` | Entero | `18_000_000` |
| `float` | Decimal | `4.7` |
| `str` | Texto (*string*) | `"Ecuador"` |
| `bool` | Lógico | `True`, `False` |

La función `type()` revela el tipo de cualquier valor.

In [5]:
# 🧑‍🏫 Yo hago — Variables para describir un país
pais = "Ecuador"                # str: texto, siempre entre comillas
poblacion = 18_000_000          # int: los guiones bajos solo mejoran la lectura
superficie_km2 = 256_370        # int
es_sudamericano = True          # bool: True o False (con mayúscula inicial)

densidad = poblacion / superficie_km2   # el resultado de una división es float

print(type(pais), type(poblacion), type(densidad), type(es_sudamericano))

# f-string: la forma moderna de combinar texto y variables → f"...{variable}..."
print(f"{pais} tiene {poblacion:,} habitantes y una densidad de {densidad:.1f} hab/km².")

<class 'str'> <class 'int'> <class 'float'> <class 'bool'>
Ecuador tiene 18,000,000 habitantes y una densidad de 70.2 hab/km².


En el `f-string` anterior, `{poblacion:,}` agrega separadores de miles y `{densidad:.1f}` redondea a 1 decimal.

**Conversión de tipos.** En la práctica los datos suelen llegar como texto (formularios, archivos CSV). Para operar con ellos hay que convertirlos explícitamente con `int()`, `float()` y `str()`.

In [6]:
# Conversión explícita de tipos
anio_texto = "2026"          # esto es texto, no un número: no se puede sumar
anio = int(anio_texto)       # ahora sí es un entero
print(anio + 4)

pib_texto = "121000.5"
pib_millones = float(pib_texto)
print(pib_millones * 2)

# De número a texto (para concatenar con +)
codigo = "EC-" + str(593)
print(codigo)

# round() controla los decimales de un float
print(round(3.14159, 2))

2030
242001.0
EC-593
3.14


### 🔨 Tú haces 2

1. Crea tres variables sobre tu provincia: `provincia` (str), `poblacion_provincia` (int, cifra aproximada) y `es_costera` (bool).
2. Calcula `porcentaje`: qué proporción de la población nacional (18 000 000) representa, en porcentaje.
3. Imprime con un f-string, con 1 decimal: `Pichincha concentra el 17.8% de la población nacional.`

In [7]:
# Tu código aquí
# 1) provincia = ..., poblacion_provincia = ..., es_costera = ...
# 2) porcentaje = ...
# 3) print(f"...")

#### ✅ Solución

In [8]:
provincia = "Pichincha"
poblacion_provincia = 3_200_000
es_costera = False

porcentaje = poblacion_provincia / 18_000_000 * 100
print(f"{provincia} concentra el {porcentaje:.1f}% de la población nacional.")

Pichincha concentra el 17.8% de la población nacional.


## 3. Operadores (15 min)

**Aritméticos:**

| Operador | Operación | Ejemplo | Resultado |
|---|---|---|---|
| `+` `-` `*` | Suma, resta, multiplicación | `3 * 4` | `12` |
| `/` | División (siempre devuelve `float`) | `7 / 2` | `3.5` |
| `//` | División entera | `7 // 2` | `3` |
| `%` | Residuo (módulo) | `7 % 2` | `1` |
| `**` | Potencia | `2 ** 3` | `8` |

**De comparación** (el resultado siempre es un `bool`): `==` igual, `!=` distinto, `<`, `>`, `<=`, `>=`. Cuidado: `==` compara, `=` asigna.

**Lógicos** (combinan condiciones): `and`, `or`, `not`.

In [9]:
# 🧑‍🏫 Yo hago — Operadores aritméticos en un caso económico
pib = 121_000_000_000        # PIB aproximado, en USD
poblacion = 18_000_000

pib_pc = pib / poblacion
print(f"PIB per cápita: {pib_pc:,.0f} USD")

# División entera y residuo: repartir 127 estudiantes en grupos de 25
estudiantes = 127
print("Grupos completos:", estudiantes // 25)
print("Estudiantes sobrantes:", estudiantes % 25)

# Potencia: PIB proyectado con crecimiento del 3% anual durante 5 años
pib_futuro = pib * (1 + 0.03) ** 5
print(f"PIB en 5 años: {pib_futuro:,.0f} USD")

PIB per cápita: 6,722 USD
Grupos completos: 5
Estudiantes sobrantes: 2
PIB en 5 años: 140,272,162,990 USD


In [10]:
# 🧑‍🏫 Yo hago — Comparación y lógicos
print(pib_pc > 10_000)          # False
print(poblacion == 18_000_000)  # True  (== compara, = asigna)
print(pib_pc != 0)              # True

# and: ambas condiciones deben cumplirse
es_ingreso_medio = pib_pc > 4_500 and pib_pc < 14_000
print("¿Ingreso medio-alto?", es_ingreso_medio)

# El algoritmo del Tú haces 1, ahora en código:
numero = 7
es_par = numero % 2 == 0
print(f"¿{numero} es par? {es_par}")
print(f"¿{numero} es impar? {not es_par}")

False
True
True
¿Ingreso medio-alto? True
¿7 es par? False
¿7 es impar? True


### 🔨 Tú haces 3

Con los datos de la celda siguiente:

1. Calcula `tasa`: la tasa de crecimiento del PIB entre 2024 y 2025, en porcentaje. Recuerda: $\text{tasa} = \left(\frac{PIB_{2025}}{PIB_{2024}} - 1\right) \times 100$
2. Crea `supera_el_2` (bool): ¿la tasa supera el 2%?
3. Imprime ambos resultados con f-strings (la tasa con 2 decimales).

In [11]:
pib_2024 = 118_000_000_000
pib_2025 = 121_000_000_000

# Tu código aquí
# 1) tasa = ...
# 2) supera_el_2 = ...
# 3) print(...)

#### ✅ Solución

In [12]:
tasa = (pib_2025 / pib_2024 - 1) * 100
supera_el_2 = tasa > 2

print(f"La economía creció {tasa:.2f}% en 2025.")
print(f"¿Superó el 2%? {supera_el_2}")

La economía creció 2.54% en 2025.
¿Superó el 2%? True


## 4. Estructuras de datos básicas (20 min)

Hasta ahora cada variable guarda **un** valor. Las estructuras de datos guardan **colecciones**.

**Listas** (`[ ]`): colecciones **ordenadas y modificables**. Regla de oro: **el índice empieza en 0**, no en 1.

In [13]:
# 🧑‍🏫 Yo hago — Listas
provincias = ["Pichincha", "Guayas", "Azuay", "Manabí"]

print(provincias[0])       # primer elemento (índice 0)
print(provincias[-1])      # último elemento (índices negativos cuentan desde el final)
print(provincias[1:3])     # rebanada: desde el índice 1 hasta ANTES del 3

provincias.append("Loja")  # agregar al final
print(provincias)

print("Elementos en la lista:", len(provincias))
print("¿Está Guayas?", "Guayas" in provincias)

Pichincha
Manabí
['Guayas', 'Azuay']
['Pichincha', 'Guayas', 'Azuay', 'Manabí', 'Loja']
Elementos en la lista: 5
¿Está Guayas? True


**Tuplas** (`( )`): como las listas, pero **inmutables** (no se pueden modificar tras crearse). Útiles para datos que no deben cambiar: coordenadas, fechas, pares fijos.

**Diccionarios** (`{ }`): pares **clave → valor**. Ideales para datos etiquetados: en lugar de recordar posiciones, accedes por nombre.

In [14]:
# 🧑‍🏫 Yo hago — Tuplas y diccionarios
coordenadas_quito = (-0.1807, -78.4678)   # (latitud, longitud)
print("Latitud de Quito:", coordenadas_quito[0])

poblacion_provincial = {
    "Pichincha": 3_200_000,
    "Guayas": 4_500_000,
    "Azuay": 900_000,
}   # cifras aproximadas, con fines didácticos

print(poblacion_provincial["Guayas"])        # acceso por clave, no por posición
poblacion_provincial["Manabí"] = 1_600_000   # agregar un par nuevo

print(list(poblacion_provincial.keys()))     # todas las claves
print(len(poblacion_provincial), "provincias registradas")

Latitud de Quito: -0.1807
4500000
['Pichincha', 'Guayas', 'Azuay', 'Manabí']
4 provincias registradas


### 🔨 Tú haces 4

1. Crea una lista `disciplinas` con 4 disciplinas de ciencias sociales e imprime la **primera** y la **última**.
2. Crea un diccionario `capitales` con 3 pares país → capital, imprime la capital de uno de ellos y **agrega** un cuarto país.

In [15]:
# Tu código aquí
# 1) disciplinas = [...]
# 2) capitales = {...}

#### ✅ Solución

In [16]:
disciplinas = ["Economía", "Sociología", "Ciencia Política", "Antropología"]
print("Primera:", disciplinas[0])
print("Última:", disciplinas[-1])

capitales = {"Ecuador": "Quito", "Colombia": "Bogotá", "Perú": "Lima"}
print("La capital de Perú es", capitales["Perú"])

capitales["Bolivia"] = "La Paz"
print(capitales)

Primera: Economía
Última: Antropología
La capital de Perú es Lima
{'Ecuador': 'Quito', 'Colombia': 'Bogotá', 'Perú': 'Lima', 'Bolivia': 'La Paz'}


## 5. Estructuras de control (25 min)

Hasta aquí el código se ejecuta línea por línea, siempre igual. Las estructuras de control permiten **decidir** (`if`) y **repetir** (`for`, `while`).

### 5.1 Condicionales: `if` / `elif` / `else`

```
if condición_1:
    # se ejecuta si condición_1 es True
elif condición_2:
    # se ejecuta si condición_1 fue False y condición_2 es True
else:
    # se ejecuta si ninguna condición anterior fue True
```

Dos detalles de sintaxis que causan la mayoría de errores del día: los **dos puntos `:`** al final de la línea y la **indentación de 4 espacios**, que en Python define qué líneas pertenecen al bloque.

In [17]:
# 🧑‍🏫 Yo hago — Clasificar un país por ingreso (umbrales referenciales del Banco Mundial)
pib_pc = 6_700   # PIB per cápita de Ecuador, USD aprox.

if pib_pc <= 1_100:
    categoria = "ingreso bajo"
elif pib_pc <= 4_500:
    categoria = "ingreso medio-bajo"
elif pib_pc <= 14_000:
    categoria = "ingreso medio-alto"
else:
    categoria = "ingreso alto"

print(f"Con {pib_pc:,} USD per cápita, el país se clasifica como {categoria}.")

Con 6,700 USD per cápita, el país se clasifica como ingreso medio-alto.


### 5.2 Bucles `for`: repetir para cada elemento

El `for` recorre una colección y ejecuta el bloque **una vez por elemento**. Con `range(a, b)` genera la secuencia de enteros desde `a` hasta `b - 1`.

In [18]:
# 🧑‍🏫 Yo hago — for sobre listas, rangos y diccionarios
for provincia in provincias:
    print("Provincia:", provincia)

# Patrón acumulador: sumar 1 + 2 + ... + 10
suma = 0
for i in range(1, 11):
    suma = suma + i
print("Suma de 1 a 10:", suma)

# Recorrer un diccionario obteniendo clave y valor a la vez
total = 0
for provincia, habitantes in poblacion_provincial.items():
    print(f"{provincia}: {habitantes:,} habitantes")
    total += habitantes   # abreviatura de: total = total + habitantes
print(f"Población total registrada: {total:,}")

Provincia: Pichincha
Provincia: Guayas
Provincia: Azuay
Provincia: Manabí
Provincia: Loja
Suma de 1 a 10: 55
Pichincha: 3,200,000 habitantes
Guayas: 4,500,000 habitantes
Azuay: 900,000 habitantes
Manabí: 1,600,000 habitantes
Población total registrada: 10,200,000


### 5.3 Bucles `while`: repetir mientras se cumpla una condición

Se usa cuando **no sabemos de antemano cuántas repeticiones** hacen falta, solo la condición de parada. Advertencia: si dentro del bucle nada cambia la condición, el bucle **nunca termina** (bucle infinito). Si te ocurre, detén la ejecución con el botón ⏹ (*Interrupt*) de VS Code.

In [19]:
# 🧑‍🏫 Yo hago — while: ¿en cuántos años se duplica el PIB creciendo al 3% anual?
pib_actual = 121.0            # miles de millones de USD
objetivo = 2 * pib_actual
anios = 0

while pib_actual < objetivo:
    pib_actual = pib_actual * 1.03   # crece 3% cada año
    anios += 1                       # si olvidas esta línea: ¡bucle infinito!

print(f"El PIB se duplica en {anios} años (la 'regla del 70' predice ≈ 70/3 ≈ {70/3:.0f}).")

El PIB se duplica en 24 años (la 'regla del 70' predice ≈ 70/3 ≈ 23).


### 🔨 Tú haces 5

Con la lista `notas` de la celda siguiente:

1. Usa un `for` con el **patrón acumulador** para calcular el promedio (sin usar `sum()`).
2. Con `if`/`elif`/`else`, clasifica el promedio: ≥ 9 → `"Excelente"`, ≥ 7 → `"Aprobado"`, en otro caso → `"En recuperación"`.
3. *(Opcional)* Con un `while`, imprime una cuenta regresiva de 5 a 1 y al final `"¡Inicio!"`.

In [20]:
notas = [7.5, 9.0, 8.0, 6.5, 10.0]

# Tu código aquí
# 1) for ... (acumulador)
# 2) if / elif / else
# 3) while ... (opcional)

#### ✅ Solución

In [21]:
# (1) Promedio con acumulador
suma = 0
for nota in notas:
    suma += nota
promedio = suma / len(notas)
print(f"Promedio: {promedio}")

# (2) Clasificación
if promedio >= 9:
    print("Excelente")
elif promedio >= 7:
    print("Aprobado")
else:
    print("En recuperación")

# (3) Cuenta regresiva (opcional)
contador = 5
while contador >= 1:
    print(contador)
    contador -= 1
print("¡Inicio!")

Promedio: 8.2
Aprobado
5
4
3
2
1
¡Inicio!


## 6. Funciones (15 min)

Una **función** empaqueta un bloque de código con nombre propio para **reutilizarlo** sin copiar y pegar. Anatomía:

```
def nombre_funcion(parametro_1, parametro_2):
    """Docstring: qué hace la función."""
    resultado = ...       # cuerpo (indentado)
    return resultado      # valor que la función devuelve
```

Distinción clave para principiantes: `print()` solo **muestra** en pantalla; `return` **devuelve** un valor que puede guardarse en una variable y seguir usándose. Una función sin `return` devuelve `None`.

In [22]:
# 🧑‍🏫 Yo hago — Definir y reutilizar funciones
def pib_per_capita(pib, poblacion):
    """Calcula el PIB per cápita en USD."""
    return pib / poblacion

def clasificar_ingreso(pib_pc):
    """Clasifica un país según su PIB per cápita (umbrales referenciales)."""
    if pib_pc <= 1_100:
        return "ingreso bajo"
    elif pib_pc <= 4_500:
        return "ingreso medio-bajo"
    elif pib_pc <= 14_000:
        return "ingreso medio-alto"
    else:
        return "ingreso alto"

# Una vez definidas, se reutilizan cuantas veces haga falta
ecuador = pib_per_capita(121_000_000_000, 18_000_000)
print(f"Ecuador: {ecuador:,.0f} USD → {clasificar_ingreso(ecuador)}")

chile = pib_per_capita(335_000_000_000, 20_000_000)
print(f"Chile:   {chile:,.0f} USD → {clasificar_ingreso(chile)}")

Ecuador: 6,722 USD → ingreso medio-alto
Chile:   16,750 USD → ingreso alto


### 🔨 Tú haces 6

1. Define `promedio(lista)` que **devuelva** el promedio de una lista de números (reutiliza tu acumulador del bloque 5).
2. Define `es_par(n)` que devuelva `True` si `n` es par y `False` si es impar (una sola línea con `return`).
3. Prueba ambas: `promedio(notas)` debe dar `8.2`; `es_par(10)` → `True`, `es_par(7)` → `False`.

In [23]:
# Tu código aquí
# def promedio(lista):
#     ...
# def es_par(n):
#     ...

#### ✅ Solución

In [24]:
def promedio(lista):
    """Devuelve el promedio de una lista de números."""
    suma = 0
    for valor in lista:
        suma += valor
    return suma / len(lista)

def es_par(n):
    """Devuelve True si n es par, False si es impar."""
    return n % 2 == 0

print(promedio(notas))        # 8.2
print(es_par(10), es_par(7))  # True False

8.2
True False


## 7. Ejercicio integrador (10 min · si falta tiempo, queda como tarea)

Combina todo lo aprendido: diccionarios, tuplas, funciones, `for`, `if` y f-strings.

Con los datos de la celda siguiente:

1. Define una función `ficha(pais, pib, poblacion)` que **devuelva** un texto con el formato:
   `Ecuador | PIB pc: 6,722 USD | ingreso medio-alto`
   Pista: reutiliza `pib_per_capita()` y `clasificar_ingreso()` del bloque 6.
2. Recorre el diccionario con un `for` e imprime la ficha de cada país.

In [25]:
# Datos: país → (PIB en USD, población). Cifras aproximadas, con fines didácticos.
datos = {
    "Ecuador":  (121_000_000_000, 18_000_000),
    "Colombia": (420_000_000_000, 52_000_000),
    "Perú":     (285_000_000_000, 34_000_000),
    "Bolivia":  ( 48_000_000_000, 12_500_000),
}

# Tu código aquí
# 1) def ficha(pais, pib, poblacion): ...
# 2) for ... : print(ficha(...))

#### ✅ Solución

In [26]:
def ficha(pais, pib, poblacion):
    pib_pc = pib_per_capita(pib, poblacion)
    return f"{pais} | PIB pc: {pib_pc:,.0f} USD | {clasificar_ingreso(pib_pc)}"

for pais in datos:
    pib, poblacion = datos[pais]      # desempaquetar la tupla en dos variables
    print(ficha(pais, pib, poblacion))

Ecuador | PIB pc: 6,722 USD | ingreso medio-alto
Colombia | PIB pc: 8,077 USD | ingreso medio-alto
Perú | PIB pc: 8,382 USD | ingreso medio-alto
Bolivia | PIB pc: 3,840 USD | ingreso medio-bajo


Variante más compacta, desempaquetando directamente en el `for`:

In [27]:
for pais, (pib, poblacion) in datos.items():
    print(ficha(pais, pib, poblacion))

Ecuador | PIB pc: 6,722 USD | ingreso medio-alto
Colombia | PIB pc: 8,077 USD | ingreso medio-alto
Perú | PIB pc: 8,382 USD | ingreso medio-alto
Bolivia | PIB pc: 3,840 USD | ingreso medio-bajo


## Resumen y referencia rápida

| Concepto | Sintaxis | Ejemplo |
|---|---|---|
| Variable | `nombre = valor` | `poblacion = 18_000_000` |
| Tipos básicos | `int`, `float`, `str`, `bool` | `type(4.7)` → `float` |
| f-string | `f"texto {variable}"` | `f"Tasa: {tasa:.2f}%"` |
| Aritméticos | `+ - * / // % **` | `7 % 2` → `1` |
| Comparación | `== != < > <= >=` | `pib_pc > 10_000` |
| Lógicos | `and`, `or`, `not` | `a > 0 and b > 0` |
| Lista | `[a, b, c]` | `provincias[0]`, `.append()` |
| Tupla | `(a, b)` | `coordenadas[0]` |
| Diccionario | `{clave: valor}` | `capitales["Ecuador"]` |
| Condicional | `if / elif / else:` | clasificar ingreso |
| Bucle definido | `for x in coleccion:` | patrón acumulador |
| Bucle condicional | `while condicion:` | años para duplicar el PIB |
| Función | `def f(x): return ...` | `pib_per_capita(pib, pob)` |

**Errores frecuentes y cómo leerlos:**

| Mensaje | Causa típica | Solución |
|---|---|---|
| `SyntaxError` | Faltan `:` al final de `if`/`for`/`def`, o comillas sin cerrar | Revisar la línea señalada |
| `IndentationError` | Bloque mal indentado | Usar siempre 4 espacios |
| `NameError` | Variable no definida (o celda anterior sin ejecutar) | Ejecutar las celdas en orden |
| `TypeError` | Mezclar tipos: `"2026" + 4` | Convertir con `int()`, `float()`, `str()` |
| `KeyError` / `IndexError` | Clave inexistente / índice fuera de rango | Verificar con `in` o `len()` |

**Para practicar en casa:** modifica los datos de los ejercicios con cifras de tu interés (tu provincia, otros países) y vuelve a ejecutar todo el notebook de arriba hacia abajo (**Run All**). Si todo corre sin errores, dominas lo esencial de la sesión.